### PR TITLE
Show retaining paths for the class instances.

### PR DESCRIPTION
**Pull request by polina-c**
https://github.com/flutter/devtools/issues/3949

https://user-images.githubusercontent.com/12115586/191857406-c18028bb-bd58-4fe1-8a3e-70602cf2f0f2.mov

### CODE REVIEW

File: packages/devtools_app/lib/src/screens/memory/panes/diff/controller/heap_diff.dart [hunk 1]
```diff
@@ -76,26 +76,15 @@ class HeapComparison {
       final youngerRecord = younger[key];
 
       if (olderRecord != null && youngerRecord != null) {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/controller/heap_diff.dart:78]: now that the class has been renamed, should all vars with 'record' in their name be changed to 'stats'?

File: packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/class_details.dart [hunk 2]
```diff
@@ -4,21 +4,53 @@
...
   Widget build(BuildContext context) {
-    final theClass = heapClass;
-    if (theClass == null) {
-      return const Center(
-        child: Text('Select class to see details here.'),
-      );
-    }
-    return Center(child: Text('Details for ${theClass.fullName} will be here'));
+    return ValueListenableBuilder<HeapClassStatistics?>(
+      valueListenable: item.selectedRecord,
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/class_details.dart:25]: should we rename selectedRecord to selectedClass or selectedClassStats?

File: packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/class_stats_table.dart [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        ..columnIndex = _columns.indexOf(_shallowSizeColumn)
+        ..initialized = true;
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return FlatTable<_RetainingPathRecord>(
+      columns: _columns,
+      data: widget.data.byRetainingPath.entries.toList(growable: false),
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/class_stats_table.dart:153]: do this in initState (and didUpdateWidget) so we don't have to copy a list on every build.

File: packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/snapshot_stats_table.dart [hunk 4]
```diff
@@ -23,27 +24,28 @@ class _ClassNameColumn extends ColumnData<HeapStatsRecord> {
         );
 
   @override
-  String? getValue(HeapStatsRecord record) => record.heapClass.className;
+  String? getValue(HeapClassStatistics record) => record.heapClass.className;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/snapshot_stats_table.dart:27]: r/record/classStats - here and elsewhere now that the class name has changed from HeapStatsRecord to HeapClassStatistics

File: packages/devtools_app/lib/src/screens/memory/shared/heap/heap.dart [hunk 5]
```diff
@@ -13,50 +13,138 @@ class AdaptedHeap {
...
 }
 
 class HeapStatistics {
   HeapStatistics(this.recordsByClass);
 
   /// Maps full class name to stats record of this class.
-  final Map<String, HeapStatsRecord> recordsByClass;
-  late final List<HeapStatsRecord> records =
+  final Map<String, HeapClassStatistics> recordsByClass;
+  late final List<HeapClassStatistics> records =
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/shared/heap/heap.dart:43]: r/recordsByClass/classStatsByClassName or statsByClassName
r/records/classStats

File: packages/devtools_app/lib/src/screens/memory/shared/heap/heap.dart [hunk 6]
```diff
@@ -13,50 +13,138 @@ class AdaptedHeap {
...
+  )   : assert(left.heapClass.fullName == right.heapClass.fullName),
+        _isSealed = true,
+        heapClass = left.heapClass,
+        total = SizeOfClassSet.subtract(left.total, right.total),
+        // ignore: dead_code
+        byRetainingPath = throw UnimplementedError();
 
   final HeapClass heapClass;
+  final SizeOfClassSet total;
+  final Map<ClassOnlyHeapPath, SizeOfClassSet> byRetainingPath;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/shared/heap/heap.dart:81]: rename to sizeByRetainingPath so it is clear what this map holds

File: packages/devtools_app/lib/src/screens/memory/shared/heap/model.dart [hunk 7]
```diff
@@ -76,41 +77,69 @@ class AdaptedHeapData {
...
-  }
+  String? shortPath() => '/${objects.map((o) => o.shortName).join('/')}/';
 
   /// Retaining path for the object as an array of the retaining objects.
-  List<String>? detailedPath(IdentityHashCode code) {
-    final path = _retainingPath(code);
-    if (path == null) return null;
-    return path.map((i) => objects[i].name).toList();
-  }
+  List<String>? detailedPath() =>
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/shared/heap/model.dart:118]: do we have the source locations for each part of this path? if so, would be great to have each part of the path be a link that will open that line of code in the IDE or in the debugger code view. Not for this PR, but an idea for the future.

File: packages/devtools_app/lib/src/screens/memory/shared/heap/model.dart [hunk 8]
```diff
@@ -76,41 +77,69 @@ class AdaptedHeapData {
...
+class ClassOnlyHeapPath {
+  ClassOnlyHeapPath(HeapPath heapPath)
+      : classes =
+            heapPath.objects.map((o) => o.heapClass).toList(growable: false);
+  final List<HeapClass> classes;
 
-/// Sequence of ids of objects in the heap.
-typedef HeapPath = List<int>;
+  String asShortString() => classes.map((e) => e.className).join('/');
+  String asLongString() => classes.map((e) => e.fullName).join('\n');
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/shared/heap/model.dart:130]: nit: maybe `asMultiLineString` instead

File: packages/devtools_app/lib/src/screens/memory/shared/heap/model.dart [hunk 9]
```diff
@@ -76,41 +77,69 @@ class AdaptedHeapData {
...
+/// Heap path represented by classes only, without object details.
+class ClassOnlyHeapPath {
+  ClassOnlyHeapPath(HeapPath heapPath)
+      : classes =
+            heapPath.objects.map((o) => o.heapClass).toList(growable: false);
+  final List<HeapClass> classes;
 
-/// Sequence of ids of objects in the heap.
-typedef HeapPath = List<int>;
+  String asShortString() => classes.map((e) => e.className).join('/');
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/shared/heap/model.dart:129]: new line after

**Review by kenzieschmoll — COMMENTED** [review]
generally looks good. can we add some widget tests?

File: packages/devtools_app/lib/src/screens/memory/shared/heap/model.dart (hunk 7 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/memory/shared/heap/model.dart:118]: It would be nice, but heap snapshot does not seem to contain location.

**Comment by polina-c** [comment]
> generally looks good. can we add some widget tests?

Is it ok to add them when we stabilize the feature?

File: packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart [hunk 10]
```diff
@@ -43,6 +44,10 @@ class DiffPaneController {
   /// informational item.
   bool get hasSnapshots => snapshots.value.length > 1;
 
+  var snapshotStatsSorting = ColumnSorting();
```
Comment (CoderDake) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/controller/diff_pane_controller.dart:47]: Should this be final?

File: packages/devtools_app/lib/src/screens/memory/shared/heap/heap.dart [hunk 11]
```diff
@@ -13,50 +13,138 @@ class AdaptedHeap {
...
+        heapClass = other.heapClass,
+        total = SizeOfClassSet.negative(other.total),
+        sizeByRetainingPath = other.sizeByRetainingPath
+            .map((key, value) => MapEntry(key, SizeOfClassSet.negative(value)));
+
+  HeapClassStatistics.subtract(
+    HeapClassStatistics left,
+    HeapClassStatistics right,
+  )   : assert(left.heapClass.fullName == right.heapClass.fullName),
+        _isSealed = true,
```
Comment (CoderDake) [inline packages/devtools_app/lib/src/screens/memory/shared/heap/heap.dart:73]: Is sealing used to make sure that we never count instances that have been sealed?
If so could we add some JSDoc explaining what the sealing is needed for or how it is being used?

**Review by CoderDake — COMMENTED** [review]
I've got to pop out for the weekend but will take a longer look at this on monday morning :D
Have a great weekend!

File: packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/class_stats_table.dart [hunk 12]
```diff
@@ -124,11 +125,14 @@ class ClassStatsTable extends StatefulWidget {
 class _ClassStatsTableState extends State<ClassStatsTable>
     with AutoDisposeMixin {
   late final List<ColumnData<_RetainingPathRecord>> _columns;
+  late final List<MapEntry<ClassOnlyHeapPath, SizeOfClassSet>> _dataList;
 
   @override
   void initState() {
     super.initState();
 
+    _dataList = widget.data.sizeByRetainingPath.entries.toList(growable: false);
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/class_stats_table.dart:134]: _dataList needs to be re-initialized in didUpdate widget when widget.data has changed

**Review by kenzieschmoll — APPROVED** [review]
lgtm once the change to didUpdateWidget has been made

File: packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/class_stats_table.dart (hunk 12 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/class_stats_table.dart]: fixed, but i am not sure i understand it

Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/class_stats_table.dart]: My comment was for `didUpdateWidget`, not `didChangeDependencies`. For a stateful widget, `didUpdateWidget` is called when the widget is rebuilt, so if the widget was rebuilt with a new value for `data`, the state needs to be updated to reflect the changed value. A good rule of thumb is any time you are doing some sort of data initialization in `initState` or `didChangeDependencies` based on the widget's values, the same initialization will need to take place in `didUpdateWidget` to ensure the state is updated properly when the widget changes.

**Pull request merged**